capabilityreceiver/pkg/capabilityhandler/validation/r4validation.go

```diff
@@ -282,186 +282,41 @@ func (v *r4Validation) KindValid(capStat capabilityparser.CapabilityStatement) [
-	baseComment := "Messaging end-point is required (and is only permitted) when a statement is for an implementation. This endpoint must be an implementation."
-	ruleError := endpointmanager.Rule{
-		RuleName:  endpointmanager.MessagingEndptRule,
-		Valid:     false,
-		Expected:  "true",
-		Actual:    "false",
-		Comment:   baseComment,
-		Reference: "http://hl7.org/fhir/capabilitystatement.html",
-		ImplGuide: "USCore 3.1",
-	}
-
-	kindRule := v.baseVal.KindValid(capStat)
-	if !kindRule[0].Valid {
-		ruleError.Comment = kindRule[0].Comment + " " + baseComment
-		return ruleError
-	}
-	messaging, err := capStat.GetMessaging()
-	if err != nil || len(messaging) == 0 {
-		ruleError.Comment = "Messaging does not exist. " + baseComment
-		return ruleError
-	}
-	for _, message := range messaging {
-		endpoints, err := capStat.GetMessagingEndpoint(message)
-		if err != nil || len(endpoints) == 0 {
-			ruleError.Comment = "Endpoint field in Messaging does not exist. " + baseComment
-			return ruleError
-		}
-	}
-
-	ruleError.Valid = true
-	ruleError.Actual = "true"
-	return ruleError
+	baseRule := v.baseVal.MessagingEndpointValid(capStat)
+	baseRule.Reference = "http://hl7.org/fhir/capabilitystatement.html"
+	return baseRule
 }
 
 // EndpointFunctionValid checks the requirement "A Capability Statement SHALL have at least one of REST,
 // messaging or document element."
 func (v *r4Validation) EndpointFunctionValid(capStat capabilityparser.CapabilityStatement) endpointmanager.Rule {
-	var actualVal []string
-	baseComment := "A Capability Statement SHALL have at least one of REST, messaging or document element."
-	ruleError := endpointmanager.Rule{
-		RuleName:  endpointmanager.EndptFunctionRule,
-		Valid:     true,
-		Expected:  "rest,messaging,document",
-		Comment:   baseComment,
-		Reference: "http://hl7.org/fhir/capabilitystatement.html",
-		ImplGuide: "USCore 3.1",
-	}
-	// If rest is not nil, add to actual list
-	rest, err := capStat.GetRest()
-	if err == nil && len(rest) > 0 {
-		actualVal = append(actualVal, "rest")
-	}
-	// If messaging is not nil, add to actual list
-	messaging, err := capStat.GetMessaging()
-	if err == nil && len(messaging) > 0 {
-		actualVal = append(actualVal, "messaging")
-	}
-	// if document is not nil, add to actual list
-	document, err := capStat.GetDocument()
-	if err == nil && len(document) > 0 {
-		actualVal = append(actualVal, "document")
-	}
-	// If none of the above exist, the capability statement is not valid
-	if len(actualVal) == 0 {
-		ruleError.Actual = ""
-		ruleError.Valid = false
-		return ruleError
-	}
-	ruleError.Actual = strings.Join(actualVal, ",")
-	return ruleError
+	baseRule := v.baseVal.EndpointFunctionValid(capStat)
+	baseRule.Reference = "http://hl7.org/fhir/capabilitystatement.html"
+	baseRule.Comment = "A Capability Statement SHALL have at least one of REST, messaging or document element."
```

Comment:
Here and on line 304, I think this could cause problems, since in the original function, if the capability/conformance statement doesn't exist, then the `baseRule.Comment` would be set to a string saying it doesn't exist, and then here it would be overwritten to the base comment. The fact that comments should say "Conformance Resource" for DSTU2 and "Capability Statement" for R4 is annoying... Maybe instead of being specific, we just say "This resource", like "This resource SHALL have at least one of rest, messaging or document element"? I think that would still be accurate but would also allow for consistency across versions?